### ISSUE TITLE
Logcollector sends old contents of the scanned file.

### ISSUE THREAD

**Issue by JcabreraC** [wazuh/wazuh#4030]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 3.10.3 | Logcollector | Agent | Sources | Windows 7 |

Enabling in `ossec.conf` of the agent `localfile` and `sca`, when starting the agent, reads some previous data from the file scanned by `localfile`.

## Example:
`ossec.conf`
```xml
<localfile>
    <location>C:\logs\test.log</location>
    <log_format>syslog</log_format>
  </localfile>

. . . 

<sca>
    <enabled>yes</enabled>
    <scan_on_start>yes</scan_on_start>
    <interval>12h</interval>
    <skip_nfs>yes</skip_nfs>
  </sca>
```
File `test.log` has the alphabet.

`archives.log` when the agent is started
```
2019 Sep 30 15:40:02 (vm-win7) 10.0.0.7->\logs\test.log H
2019 Sep 30 15:40:02 (vm-win7) 10.0.0.7->\logs\test.log I
2019 Sep 30 15:40:02 (vm-win7) 10.0.0.7->\logs\test.log J
2019 Sep 30 15:40:02 (vm-win7) 10.0.0.7->\logs\test.log K
2019 Sep 30 15:40:02 (vm-win7) 10.0.0.7->\logs\test.log L
2019 Sep 30 15:40:02 (vm-win7) 10.0.0.7->\logs\test.log M
2019 Sep 30 15:40:02 (vm-win7) 10.0.0.7->\logs\test.log N
2019 Sep 30 15:40:02 (vm-win7) 10.0.0.7->\logs\test.log O
2019 Sep 30 15:40:02 (vm-win7) 10.0.0.7->\logs\test.log P
2019 Sep 30 15:40:02 (vm-win7) 10.0.0.7->\logs\test.log Q
2019 Sep 30 15:40:02 (vm-win7) 10.0.0.7->\logs\test.log R
2019 Sep 30 15:40:02 (vm-win7) 10.0.0.7->\logs\test.log S
2019 Sep 30 15:40:02 (vm-win7) 10.0.0.7->\logs\test.log T
2019 Sep 30 15:40:02 (vm-win7) 10.0.0.7->\logs\test.log U
2019 Sep 30 15:40:02 (vm-win7) 10.0.0.7->\logs\test.log V
2019 Sep 30 15:40:02 (vm-win7) 10.0.0.7->\logs\test.log W
2019 Sep 30 15:40:02 (vm-win7) 10.0.0.7->\logs\test.log X
2019 Sep 30 15:40:02 (vm-win7) 10.0.0.7->\logs\test.log Y

```

Example [ossec.conf](https://github.com/wazuh/wazuh/files/3738631/ossec.txt)

**Comment by chemamartinez** [comment]
Take into account the explanation at https://github.com/wazuh/wazuh/issues/3546#issuecomment-503158172.

**Comment by sergiovp** [comment]
Hello team.

I could not reproduce this issue on Windows XP.

I created a file named `mytest.txt` on `C:\Test` path.
The content of that file is:
~~~
Hello.
This is a test
~~~
And following @JcabreraC steps, it worked as expected:
~~~
2019 Nov 13 10:09:04 (vela-b5e7a638c2) 10.0.0.1->ossec ossec: Agent started: 'vela-b5e7a638c2->10.0.0.1'.
2019 Nov 13 10:09:07 (vela-b5e7a638c2) 10.0.0.1->sca {"type":"summary","scan_id":375581369,"name":"Benchmark for Windows audit","policy_id":"sca_win_audit","file":"sca_win_audit.yml","description":"This document provides a way of ensuring the security of the Windows systems.","passed":19,"failed":16,"invalid":36,"total_checks":71,"score":54.285713195800781,"start_time":1573639744,"end_time":1573639744,"hash":"88dbc3e2a6b772eb1ff8a743422401b900f67837a67a414d112e9445d4908f71","hash_file":"defe15f736861cd24abf24d88d5f9b4fe420307566201de418c6c12e71b237dd","first_scan":1}
2019 Nov 13 10:09:07 (vela-b5e7a638c2) 10.0.0.1->sca {"type":"policies","policies":["sca_win_audit"]}
2019 Nov 13 10:09:18 (vela-b5e7a638c2) 10.0.0.1->rootcheck Starting rootcheck scan.
2019 Nov 13 10:09:23 (vela-b5e7a638c2) 10.0.0.1->rootcheck Ending rootcheck scan.
~~~
If I modifie the file, adding `it seems it works` I get the following:
~~~
2019 Nov 13 10:10:49 (vela-b5e7a638c2) 10.0.0.1->\Test\mytest.txt It seems it works!
~~~

I'll check if it works in others Windows OS.

**Comment by sergiovp** [comment]
## Update

Finally, I could reproduce the issue.
I created several files which have random letters.
![image](https://user-images.githubusercontent.com/36703917/68757880-ffd8e480-060c-11ea-8991-06ba968b1713.png)

But we only get reported by some files.
![image](https://user-images.githubusercontent.com/36703917/68757940-226afd80-060d-11ea-8a2b-3f8a75fca868.png)

It means that probably @crolopez is right and it happens because of what he says [here](https://github.com/wazuh/wazuh/issues/3546#issuecomment-503158172).

**Comment by sergiovp** [comment]
## Advance
I have been researching about a function which solves `_ftelli64` problems as @crolopez says [here](https://github.com/wazuh/wazuh/issues/3546#issuecomment-503158172).
Before using `_ftelli64` function, we used `ftell`, our problem was that ftell was not able to read files positions beyond than 2GiB.
It was then that I started to read about `ftello64` function. It seems the same as `_ftelli64` but, `ftello64` does't have its problem.
I'll explain it with some examples.

To reproduce the issue, you may use a configuration as follows in the Windows agent:
~~~
<localfile>
   <location>C:\Test\*</location>
   <log_format>syslog</log_format>
</localfile>
~~~
In C:\Test path I have created several files which have random characters.
Let's see what happend when we use `ftell`, `_ftelli64` and `ftello64` functions:

+ ftell.
It does't read any previous data from the file scanned by localfile.

+ _ftelli64.
It reads some previous data from the file scanned by localfile. It does it randomly.
~~~
2019 Nov 15 09:21:13 (vela-b5e7a638c2) 10.0.0.1->\Test\mytest2.txt a
2019 Nov 15 09:21:13 (vela-b5e7a638c2) 10.0.0.1->\Test\mytest2.txt a
2019 Nov 15 09:21:13 (vela-b5e7a638c2) 10.0.0.1->\Test\mytest2.txt a
2019 Nov 15 09:21:13 (vela-b5e7a638c2) 10.0.0.1->\Test\mytest2.txt a
2019 Nov 15 09:21:13 (vela-b5e7a638c2) 10.0.0.1->\Test\mytest2.txt aa
2019 Nov 15 09:21:13 (vela-b5e7a638c2) 10.0.0.1->\Test\mytest2.txt a
2019 Nov 15 09:21:13 (vela-b5e7a638c2) 10.0.0.1->\Test\mytest2.txt a
~~~

+ ftello64.
It does't read any previous data from the file scanned by localfile.

The difference between `ftell` and `ftello64` is that in contrast to ftell, `ftello64` is able to read files positions beyond than 2GiB.

This three functions are limited by files which have more than **10.001** lines.
It means that if we have a file with 10.003 lines, it will read the two last lines of our file previous content.
In terms of file size, we don't have any type of problem using `ftello64` function.
I created a file whose size is bigger than 3GiB with less than 10.000 lines.
![image](https://user-images.githubusercontent.com/36703917/68939561-3c3d4980-07a1-11ea-9701-39a5366a09fe.png)
It does't read any previous data from the file scanned by localfile and if we write something:
`@echo 123>> test.txt`
We are notified:
~~~
2019 Nov 15 11:14:14 (vela-b5e7a638c2) 10.0.0.1->\Test\sergio 123
~~~

In summary, it works as expected under the condition that we don't have a file with more than 10.001 lines, in which case It will read as many lines as the difference between the total lines and 10.001.

**Comment by sergiovp** [comment]
## Last step

Hello team.

What I commented above about the 10.000 lines is due to by default, in `internals_options.conf` file is specified: 
~~~
# Logcollector - Maximum number of lines to read from the same file [100..1000000]
# 0. Disable line burst limitation
logcollector.max_lines=10000
~~~
That is the reason why from 10.000 lines the content of the file is read.

In short, `ftello64` function works as expected.

**Comment by JcabreraC** [comment]
This issue is solved in https://github.com/wazuh/wazuh/issues/3368

Regards !
